Index: linux-2.6.27.y/kernel/ptrace.c
===================================================================
--- linux-2.6.27.y.orig/kernel/ptrace.c
+++ linux-2.6.27.y/kernel/ptrace.c
@@ -140,7 +140,7 @@ int __ptrace_may_access(struct task_stru
 		dumpable = get_dumpable(task->mm);
 	if (!dumpable && !capable(CAP_SYS_PTRACE))
 		return -EPERM;
-	if (!vx_check(task->xid, VS_ADMIN_P|VS_IDENT))
+	if (!vx_check(task->xid, VS_WATCH_P | VS_IDENT))
 		return -EPERM;
 	if (!vx_check(task->xid, VS_IDENT) &&
 		!task_vx_flags(task, VXF_STATE_ADMIN, 0))
