diff --git a/fs/ext2/super.c b/fs/ext2/super.c
index 42c0fe7..46a1d7a 100644
--- a/fs/ext2/super.c
+++ b/fs/ext2/super.c
@@ -416,6 +416,7 @@ static const match_table_t tokens = {
 	{Opt_noacl, "noacl"},
 	{Opt_xip, "xip"},
 	{Opt_tag, "tag"},
+	{Opt_tag, "tagxid"},
 	{Opt_notag, "notag"},
 	{Opt_tagid, "tagid=%u"},
 	{Opt_grpquota, "grpquota"},
diff --git a/fs/ext3/super.c b/fs/ext3/super.c
index e259a99..d02155b 100644
--- a/fs/ext3/super.c
+++ b/fs/ext3/super.c
@@ -876,6 +876,7 @@ static const match_table_t tokens = {
 	{Opt_nobarrier, "nobarrier"},
 	{Opt_resize, "resize"},
 	{Opt_tag, "tag"},
+	{Opt_tag, "tagxid"},
 	{Opt_notag, "notag"},
 	{Opt_tagid, "tagid=%u"},
 	{Opt_err, NULL},
diff --git a/fs/ext4/super.c b/fs/ext4/super.c
index 94d8c6c..90bb7c0 100644
--- a/fs/ext4/super.c
+++ b/fs/ext4/super.c
@@ -1251,6 +1251,7 @@ static const match_table_t tokens = {
 	{Opt_init_itable, "init_itable"},
 	{Opt_noinit_itable, "noinit_itable"},
 	{Opt_tag, "tag"},
+	{Opt_tag, "tagxid"},
 	{Opt_notag, "notag"},
 	{Opt_tagid, "tagid=%u"},
 	{Opt_err, NULL},
diff --git a/fs/ocfs2/super.c b/fs/ocfs2/super.c
index fa2baad..e28e02f 100644
--- a/fs/ocfs2/super.c
+++ b/fs/ocfs2/super.c
@@ -201,6 +201,7 @@ static const match_table_t tokens = {
 	{Opt_usrquota, "usrquota"},
 	{Opt_grpquota, "grpquota"},
 	{Opt_tag, "tag"},
+	{Opt_tag, "tagxid"},
 	{Opt_notag, "notag"},
 	{Opt_tagid, "tagid=%u"},
 	{Opt_err, NULL}
