diff -NurpP --exclude '*.orig' --exclude '*.rej' linux-2.6.27.10-vs2.3.x-P/fs/ext2/super.c linux-2.6.27.10-vs2.3.x-P570/fs/ext2/super.c
--- linux-2.6.27.10-vs2.3.x-P/fs/ext2/super.c	2008-10-13 14:54:20.000000000 +0200
+++ linux-2.6.27.10-vs2.3.x-P570/fs/ext2/super.c	2009-01-12 01:18:23.000000000 +0100
@@ -420,6 +420,7 @@ static match_table_t tokens = {
 	{Opt_noacl, "noacl"},
 	{Opt_xip, "xip"},
 	{Opt_tag, "tag"},
+	{Opt_tag, "tagxid"},
 	{Opt_notag, "notag"},
 	{Opt_tagid, "tagid=%u"},
 	{Opt_grpquota, "grpquota"},
diff -NurpP --exclude '*.orig' --exclude '*.rej' linux-2.6.27.10-vs2.3.x-P/fs/ext3/super.c linux-2.6.27.10-vs2.3.x-P570/fs/ext3/super.c
--- linux-2.6.27.10-vs2.3.x-P/fs/ext3/super.c	2008-11-17 18:20:41.000000000 +0100
+++ linux-2.6.27.10-vs2.3.x-P570/fs/ext3/super.c	2009-01-12 01:18:23.000000000 +0100
@@ -809,6 +809,7 @@ static match_table_t tokens = {
 	{Opt_barrier, "barrier=%u"},
 	{Opt_resize, "resize"},
 	{Opt_tag, "tag"},
+	{Opt_tag, "tagxid"},
 	{Opt_notag, "notag"},
 	{Opt_tagid, "tagid=%u"},
 	{Opt_err, NULL},
diff -NurpP --exclude '*.orig' --exclude '*.rej' linux-2.6.27.10-vs2.3.x-P/fs/ext4/super.c linux-2.6.27.10-vs2.3.x-P570/fs/ext4/super.c
--- linux-2.6.27.10-vs2.3.x-P/fs/ext4/super.c	2008-12-08 04:52:52.000000000 +0100
+++ linux-2.6.27.10-vs2.3.x-P570/fs/ext4/super.c	2009-01-12 01:18:23.000000000 +0100
@@ -969,6 +969,7 @@ static match_table_t tokens = {
 	{Opt_delalloc, "delalloc"},
 	{Opt_nodelalloc, "nodelalloc"},
 	{Opt_tag, "tag"},
+	{Opt_tag, "tagxid"},
 	{Opt_notag, "notag"},
 	{Opt_tagid, "tagid=%u"},
 	{Opt_err, NULL},
diff -NurpP --exclude '*.orig' --exclude '*.rej' linux-2.6.27.10-vs2.3.x-P/fs/ocfs2/super.c linux-2.6.27.10-vs2.3.x-P570/fs/ocfs2/super.c
--- linux-2.6.27.10-vs2.3.x-P/fs/ocfs2/super.c	2008-10-13 14:54:20.000000000 +0200
+++ linux-2.6.27.10-vs2.3.x-P570/fs/ocfs2/super.c	2009-01-12 01:18:23.000000000 +0100
@@ -175,6 +175,7 @@ static match_table_t tokens = {
 	{Opt_localflocks, "localflocks"},
 	{Opt_stack, "cluster_stack=%s"},
 	{Opt_tag, "tag"},
+	{Opt_tag, "tagxid"},
 	{Opt_notag, "notag"},
 	{Opt_tagid, "tagid=%u"},
 	{Opt_err, NULL}
