diff -Nurb linux-2.6.27.10-clickport/net/core/dev.c trellis2/net/core/dev.c
--- linux-2.6.27.10-clickport/net/core/dev.c	2009-01-27 16:02:45.000000000 -0500
+++ trellis2/net/core/dev.c	2009-02-12 11:00:46.000000000 -0500
@@ -2559,7 +2559,7 @@
 
 	total = 0;
 	for_each_netdev(net, dev) {
-		if (!nx_dev_visible(current->nx_info, dev))
+		if (net==&init_dev && !nx_dev_visible(current->nx_info, dev))
 			continue;
 		for (i = 0; i < NPROTO; i++) {
 			if (gifconf_list[i]) {
@@ -2628,8 +2628,9 @@
 static void dev_seq_printf_stats(struct seq_file *seq, struct net_device *dev)
 {
 	struct net_device_stats *stats = dev->get_stats(dev);
+    struct net *net = seq->private;
 
-	if (!nx_dev_visible(current->nx_info, dev))
+	if (net==&init_net && !nx_dev_visible(current->nx_info, dev))
 		return;
 
 	seq_printf(seq, "%6s:%8lu %7lu %4lu %4lu %4lu %5lu %10lu %9lu "
