diff -NurpP --exclude '*.orig' --exclude '*.rej' linux-2.6.27.10-vs2.3.x-P/kernel/vserver/context.c linux-2.6.27.10-vs2.3.x-P580/kernel/vserver/context.c
--- linux-2.6.27.10-vs2.3.x-P/kernel/vserver/context.c	2008-10-31 03:56:55.000000000 +0100
+++ linux-2.6.27.10-vs2.3.x-P580/kernel/vserver/context.c	2009-01-12 01:18:23.000000000 +0100
@@ -419,7 +419,7 @@ int get_xid_list(int index, unsigned int
 	int hindex, nr_xids = 0;
 
 	/* only show current and children */
-	if (!vx_check(0, VS_ADMIN | VS_WATCH)) {
+	if (!vx_check(0, VS_ADMIN | VS_WATCH) && vx_flags(VXF_HIDE_VINFO, 0)) {
 		if (index > 0)
 			return 0;
 		xids[nr_xids] = vx_current_xid();
diff -NurpP --exclude '*.orig' --exclude '*.rej' linux-2.6.27.10-vs2.3.x-P/kernel/vserver/network.c linux-2.6.27.10-vs2.3.x-P580/kernel/vserver/network.c
--- linux-2.6.27.10-vs2.3.x-P/kernel/vserver/network.c	2008-10-13 14:54:20.000000000 +0200
+++ linux-2.6.27.10-vs2.3.x-P580/kernel/vserver/network.c	2009-01-12 01:18:23.000000000 +0100
@@ -392,7 +392,7 @@ int get_nid_list(int index, unsigned int
 	int hindex, nr_nids = 0;
 
 	/* only show current and children */
-	if (!nx_check(0, VS_ADMIN | VS_WATCH)) {
+	if (!nx_check(0, VS_ADMIN | VS_WATCH) && vx_flags(VXF_HIDE_VINFO, 0)) {
 		if (index > 0)
 			return 0;
 		nids[nr_nids] = nx_current_nid();
