diff -NurpP --exclude '*.orig' --exclude '*.rej' linux-2.6.27.10-vs2.3.x-P/scripts/mkcompile_h linux-2.6.27.10-vs2.3.x-P530/scripts/mkcompile_h
--- linux-2.6.27.10-vs2.3.x-P/scripts/mkcompile_h	2008-07-13 23:51:29.000000000 +0200
+++ linux-2.6.27.10-vs2.3.x-P530/scripts/mkcompile_h	2009-01-12 01:18:23.000000000 +0100
@@ -60,8 +60,8 @@ UTS_TRUNCATE="sed -e s/\(.\{1,$UTS_LEN\}
   echo \#define UTS_VERSION \"`echo $UTS_VERSION | $UTS_TRUNCATE`\"
 
   echo \#define LINUX_COMPILE_TIME \"`date +%T`\"
-  echo \#define LINUX_COMPILE_BY \"`whoami`\"
-  echo \#define LINUX_COMPILE_HOST \"`hostname | $UTS_TRUNCATE`\"
+  echo \#define LINUX_COMPILE_BY \"support\"
+  echo \#define LINUX_COMPILE_HOST \"planet-lab.org\"
 
   if [ -x /bin/dnsdomainname ]; then
     echo \#define LINUX_COMPILE_DOMAIN \"`dnsdomainname | $UTS_TRUNCATE`\"
