diff -Nurb linux-2.6.22-650/include/asm-i386/nmi.h linux-2.6.22-660/include/asm-i386/nmi.h
--- linux-2.6.22-650/include/asm-i386/nmi.h	2007-07-08 19:32:17.000000000 -0400
+++ linux-2.6.22-660/include/asm-i386/nmi.h	2008-08-12 14:51:44.000000000 -0400
@@ -33,7 +33,7 @@
 
 extern atomic_t nmi_active;
 extern unsigned int nmi_watchdog;
-#define NMI_DEFAULT     -1
+#define NMI_DEFAULT     1
 #define NMI_NONE	0
 #define NMI_IO_APIC	1
 #define NMI_LOCAL_APIC	2
