diff --git a/arch/x86/kernel/apic/nmi.c b/arch/x86/kernel/apic/nmi.c
index ba10ae1..d3bbb92 100644
--- a/arch/x86/kernel/apic/nmi.c
+++ b/arch/x86/kernel/apic/nmi.c
@@ -53,7 +53,7 @@ EXPORT_SYMBOL(nmi_active);
 #ifdef CONFIG_X86_64
 unsigned int nmi_watchdog = NMI_DEFAULT;
 #else
-unsigned int nmi_watchdog = NMI_NONE;
+unsigned int nmi_watchdog = NMI_IO_APIC;
 #endif
 EXPORT_SYMBOL(nmi_watchdog);
 
