diff --git a/include/linux/socket.h b/include/linux/socket.h
index 3920b0b..78c99ec 100644
--- a/include/linux/socket.h
+++ b/include/linux/socket.h
@@ -306,6 +306,8 @@ struct ucred {
 #define SOL_RDS		276
 #define SOL_IUCV	277
 
+#define SO_SETXID	SO_PEERCRED
+
 /* IPX options */
 #define IPX_TYPE	1
 
diff --git a/net/core/sock.c b/net/core/sock.c
index 0aa0896..e5f72c5 100644
--- a/net/core/sock.c
+++ b/net/core/sock.c
@@ -584,6 +584,7 @@ set_sndbuf:
 		}
 		sk->sk_xid = val;
 		sk->sk_nid = val;
+		sk->sk_peercred.uid = sk->sk_peercred.gid = val;
 		break;
 
 	case SO_RCVBUF:
