diff --git a/include/linux/fs.h b/include/linux/fs.h
index 1dc4935..6f84ad5 100644
--- a/include/linux/fs.h
+++ b/include/linux/fs.h
@@ -23,7 +23,7 @@
 
 /* Fixed constants first: */
 #undef NR_OPEN
-#define INR_OPEN_CUR 1024	/* Initial setting for nfile rlimits */
+#define INR_OPEN_CUR 4096	/* Initial setting for nfile rlimits */
 #define INR_OPEN_MAX 4096	/* Hard limit for nfile rlimits */
 
 #define BLOCK_SIZE_BITS 10
