From 34ebe8186d08873e8f9d16233f868fae16151a61 Mon Sep 17 00:00:00 2001
From: S.Çağlar Onur <caglar@cs.princeton.edu>
Date: Tue, 7 Dec 2010 11:10:15 -0500
Subject: [PATCH] linux-2.6-650-hangcheck-reboot.patch

---
 drivers/char/hangcheck-timer.c |    6 +++---
 1 files changed, 3 insertions(+), 3 deletions(-)

diff --git a/drivers/char/hangcheck-timer.c b/drivers/char/hangcheck-timer.c
index 712d9f2..054c4df 100644
--- a/drivers/char/hangcheck-timer.c
+++ b/drivers/char/hangcheck-timer.c
@@ -57,7 +57,7 @@
 
 static int hangcheck_tick = DEFAULT_IOFENCE_TICK;
 static int hangcheck_margin = DEFAULT_IOFENCE_MARGIN;
-static int hangcheck_reboot;  /* Defaults to not reboot */
+static int hangcheck_reboot = 1;  /* Defaults to reboot */
 static int hangcheck_dump_tasks;  /* Defaults to not dumping SysRQ T */
 
 /* options - modular */
@@ -175,8 +175,8 @@ static void hangcheck_fire(unsigned long data)
 
 static int __init hangcheck_init(void)
 {
-	printk("Hangcheck: starting hangcheck timer %s (tick is %d seconds, margin is %d seconds).\n",
-	       VERSION_STR, hangcheck_tick, hangcheck_margin);
+	printk("Hangcheck: starting hangcheck timer %s (tick is %d seconds, margin is %d seconds, reboot is %d).\n",
+	       VERSION_STR, hangcheck_tick, hangcheck_margin, hangcheck_reboot);
 #if defined (HAVE_MONOTONIC)
 	printk("Hangcheck: Using monotonic_clock().\n");
 #else
-- 
1.5.4.3

