diff -NurpP linux-2.6.38.7-vs2.3.0.37-rc15/fs/namei.c linux-2.6.38.7-vs2.3.0.37-rc15.1/fs/namei.c
--- linux-2.6.38.7-vs2.3.0.37-rc15/fs/namei.c	2011-04-23 22:33:35.000000000 +0200
+++ linux-2.6.38.7-vs2.3.0.37-rc15.1/fs/namei.c	2011-05-29 22:41:46.000000000 +0200
@@ -2220,7 +2220,8 @@ int may_open(struct path *path, int acc_
 	}
 
 #ifdef	CONFIG_VSERVER_COWBL
-	if (IS_COW(inode) && (flag & FMODE_WRITE)) {
+	if (IS_COW(inode) && 
+		((flag & O_ACCMODE) != O_RDONLY)) {
 		if (IS_COW_LINK(inode))
 			return -EMLINK;
 		inode->i_flags &= ~(S_IXUNLINK|S_IMMUTABLE);
