diff -NurpP --exclude '*.orig' --exclude '*.rej' linux-2.6.27.10-vs2.3.x-P/arch/x86/kernel/nmi.c linux-2.6.27.10-vs2.3.x-P660/arch/x86/kernel/nmi.c
--- linux-2.6.27.10-vs2.3.x-P/arch/x86/kernel/nmi.c	2008-10-13 14:51:58.000000000 +0200
+++ linux-2.6.27.10-vs2.3.x-P660/arch/x86/kernel/nmi.c	2009-01-21 02:24:07.000000000 +0100
@@ -52,7 +52,7 @@ static cpumask_t backtrace_mask = CPU_MA
 atomic_t nmi_active = ATOMIC_INIT(0);		/* oprofile uses this */
 EXPORT_SYMBOL(nmi_active);
 
-unsigned int nmi_watchdog = NMI_NONE;
+unsigned int nmi_watchdog = NMI_IO_APIC;
 EXPORT_SYMBOL(nmi_watchdog);
 
 static int panic_on_timeout;
