diff -NurpP --exclude '*.orig' --exclude '*.rej' linux-2.6.27.10-vs2.3.x-P/kernel/vserver/Kconfig linux-2.6.27.10-vs2.3.x-P500/kernel/vserver/Kconfig
--- linux-2.6.27.10-vs2.3.x-P/kernel/vserver/Kconfig	2008-10-13 14:54:20.000000000 +0200
+++ linux-2.6.27.10-vs2.3.x-P500/kernel/vserver/Kconfig	2009-01-12 01:18:23.000000000 +0100
@@ -4,6 +4,15 @@
 
 menu "Linux VServer"
 
+config VSERVER_FILESHARING
+	bool	"(PLANETLAB) Disable Hidden File Module"
+	default y
+	help
+	  This retains the module that when a vserver can
+	  see a file, it can also stat and read it.  The
+	  assumption is that chroot() works and vservers
+	  are contained that way.
+
 config	VSERVER_AUTO_LBACK
 	bool    "Automatically Assign Loopback IP"
 	default y
