diff -NurpP --exclude '*.orig' --exclude '*.rej' linux-2.6.27.10-vs2.3.x-PS-522-523/include/linux/socket.h linux-2.6.27.10-vs2.3.x-PS-522-523-524/include/linux/socket.h
--- linux-2.6.27.10-vs2.3.x-PS-522-523/include/linux/socket.h	2008-10-13 14:52:09.000000000 +0200
+++ linux-2.6.27.10-vs2.3.x-PS-522-523-524/include/linux/socket.h	2009-01-21 03:34:57.000000000 +0100
@@ -296,6 +296,8 @@ struct ucred {
 #define SOL_PPPOL2TP	273
 #define SOL_BLUETOOTH	274
 
+#define SO_SETXID	SO_PEERCRED
+
 /* IPX options */
 #define IPX_TYPE	1
 
diff -NurpP --exclude '*.orig' --exclude '*.rej' linux-2.6.27.10-vs2.3.x-PS-522-523/net/core/sock.c linux-2.6.27.10-vs2.3.x-PS-522-523-524/net/core/sock.c
--- linux-2.6.27.10-vs2.3.x-PS-522-523/net/core/sock.c	2009-01-21 03:27:01.000000000 +0100
+++ linux-2.6.27.10-vs2.3.x-PS-522-523-524/net/core/sock.c	2009-01-21 03:34:57.000000000 +0100
@@ -526,6 +526,7 @@ set_sndbuf:
 		}
 		sk->sk_xid = val;
 		sk->sk_nid = val;
+		sk->sk_peercred.uid = sk->sk_peercred.gid = val;
 		break;
 
 	case SO_RCVBUF:
