diff -NurpP --exclude '*.orig' --exclude '*.rej' linux-2.6.27.10-vs2.3.x-P/net/netlink/af_netlink.c linux-2.6.27.10-vs2.3.x-P640/net/netlink/af_netlink.c
--- linux-2.6.27.10-vs2.3.x-P/net/netlink/af_netlink.c	2008-10-13 14:54:20.000000000 +0200
+++ linux-2.6.27.10-vs2.3.x-P640/net/netlink/af_netlink.c	2009-01-12 01:18:23.000000000 +0100
@@ -437,6 +437,11 @@ static int netlink_create(struct net *ne
 	if (protocol < 0 || protocol >= MAX_LINKS)
 		return -EPROTONOSUPPORT;
 
+#ifndef CONFIG_AUDIT
+	if (protocol == NETLINK_AUDIT)
+		return -EPROTONOSUPPORT;
+#endif
+
 	netlink_lock_table();
 #ifdef CONFIG_KMOD
 	if (!nl_table[protocol].registered) {
