diff -NurpP --exclude '*.orig' --exclude '*.rej' linux-2.6.27.10-vs2.3.x-P/include/linux/fs.h linux-2.6.27.10-vs2.3.x-P550/include/linux/fs.h
--- linux-2.6.27.10-vs2.3.x-P/include/linux/fs.h	2008-10-31 18:18:18.000000000 +0100
+++ linux-2.6.27.10-vs2.3.x-P550/include/linux/fs.h	2009-01-12 02:42:12.000000000 +0100
@@ -23,6 +23,7 @@
 #undef NR_OPEN
 extern int sysctl_nr_open;
 #define INR_OPEN 1024		/* Initial setting for nfile rlimits */
+#define INR_OPEN 4096		/* Initial setting for nfile rlimits */
 
 #define BLOCK_SIZE_BITS 10
 #define BLOCK_SIZE (1<<BLOCK_SIZE_BITS)
diff -NurpP --exclude '*.orig' --exclude '*.rej' linux-2.6.27.10-vs2.3.x-PS-02.0/include/linux/fs.h linux-2.6.27.10-vs2.3.x-PS-02.1/include/linux/fs.h
--- linux-2.6.27.10-vs2.3.x-PS-02.0/include/linux/fs.h	2009-01-25 02:29:32.000000000 +0100
+++ linux-2.6.27.10-vs2.3.x-PS-02.1/include/linux/fs.h	2009-01-21 04:26:21.000000000 +0100
@@ -22,7 +22,6 @@
 /* Fixed constants first: */
 #undef NR_OPEN
 extern int sysctl_nr_open;
-#define INR_OPEN 1024		/* Initial setting for nfile rlimits */
 #define INR_OPEN 4096		/* Initial setting for nfile rlimits */
 
 #define BLOCK_SIZE_BITS 10
