diff -NurpP --exclude '*.orig' --exclude '*.rej' linux-2.6.27.10-vs2.3.x-P/arch/x86/pci/common.c linux-2.6.27.10-vs2.3.x-P560/arch/x86/pci/common.c
--- linux-2.6.27.10-vs2.3.x-P/arch/x86/pci/common.c	2008-10-13 14:51:59.000000000 +0200
+++ linux-2.6.27.10-vs2.3.x-P560/arch/x86/pci/common.c	2009-01-15 00:02:44.000000000 +0100
@@ -17,8 +17,7 @@
 
 #include "pci.h"
 
-unsigned int pci_probe = PCI_PROBE_BIOS | PCI_PROBE_CONF1 | PCI_PROBE_CONF2 |
-				PCI_PROBE_MMCONF;
+unsigned int pci_probe = PCI_PROBE_BIOS | PCI_PROBE_CONF1 | PCI_PROBE_CONF2;
 
 unsigned int pci_early_dump_regs;
 static int pci_bf_sort;
@@ -477,6 +476,10 @@ char * __devinit  pcibios_setup(char *st
 		pci_probe |= PCI_CHECK_ENABLE_AMD_MMCONF;
 		return NULL;
 	}
+	else if (!strcmp(str, "mmconf")) {
+		pci_probe |= PCI_PROBE_MMCONF;
+		return NULL;
+	}
 #endif
 	else if (!strcmp(str, "noacpi")) {
 		acpi_noirq_set();
